File: src/state/app_state.py
"""Application state — single source of truth."""

import tomllib
from pathlib import Path

from ..core.folder import scan_folder

_STATE_FILE = "state.toml"


class AppState:
    def __init__(self, config_dir: Path | None = None):
        if config_dir is None:
            config_dir = Path.home() / ".config" / "omarchy-imageview"
        self.config_dir = Path(config_dir)

        self.current_folder: Path | None = None
        self.files: list[Path] = []
        self.index: int = 0
        self.view_mode: str = "grid"
        self.zoom_fit: bool = True
        self.zoom: float = 1.0
        self.pan_x: float = 0.0
        self.pan_y: float = 0.0
        self.last_folder: Path | None = None

    @property
    def current_file(self) -> Path | None:
        if not self.files or self.index >= len(self.files):
            return None
        return self.files[self.index]

    def load_folder(self, folder: Path, target_file: Path | None = None):
        self.current_folder = Path(folder)
        self.files = scan_folder(folder)
        self.last_folder = self.current_folder

        if target_file and target_file in self.files:
            self.index = self.files.index(target_file)
        else:
            self.index = 0

    def navigate_next(self):
        if self.files and self.index < len(self.files) - 1:
            self.index += 1

    def navigate_prev(self):
        if self.index > 0:
            self.index -= 1

    def navigate_to(self, index: int):
        if self.files:
            self.index = max(0, min(index, len(self.files) - 1))

    def remove_file(self, path: Path):
        if path in self.files:
            was_index = self.files.index(path)
            self.files.remove(path)
            if self.index >= len(self.files):
                self.index = max(0, len(self.files) - 1)
            elif self.index > was_index:
                self.index -= 1

    def save(self):
        self.config_dir.mkdir(parents=True, exist_ok=True)
        state_path = self.config_dir / _STATE_FILE
        lines = []
        if self.last_folder:
            lines.append(f'last_folder = "{self.last_folder}"')
        state_path.write_text("\n".join(lines) + "\n")

    def restore(self):
        state_path = self.config_dir / _STATE_FILE
        if not state_path.exists():
            return
        try:
            data = tomllib.loads(state_path.read_text())
            lf = data.get("last_folder")
            if lf:
                self.last_folder = Path(lf)
        except Exception:
            pass

File: src/state/app_state.rs
use std::path::{Path, PathBuf};

use crate::core::folder::SortMode;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ViewMode {
    Grid,
    Single,
}

pub struct AppState {
    pub current_folder: Option<PathBuf>,
    pub files: Vec<PathBuf>,
    pub index: usize,
    pub view_mode: ViewMode,
    pub zoom_fit: bool,
    pub zoom: f64,
    pub last_folder: Option<PathBuf>,
    /// Unsaved clockwise rotation applied on screen (0, 90, 180, 270).
    pub pending_rotation: u32,
    pub sort: SortMode,
    config_dir: PathBuf,
}

impl AppState {
    pub fn new(config_dir: Option<PathBuf>) -> Self {
        let dir = config_dir.unwrap_or_else(|| {
            dirs::config_dir()
                .unwrap_or_else(|| PathBuf::from("/tmp"))
                .join("omarchy-imageview")
        });
        Self {
            current_folder: None,
            files: Vec::new(),
            index: 0,
            view_mode: ViewMode::Grid,
            zoom_fit: false,
            zoom: 1.0,
            last_folder: None,
            pending_rotation: 0,
            sort: SortMode::Date,
            config_dir: dir,
        }
    }

    pub fn current_file(&self) -> Option<&Path> {
        self.files.get(self.index).map(|p| p.as_path())
    }

    pub fn load_folder(&mut self, folder: &Path, target_file: Option<&Path>) {
        self.current_folder = Some(folder.to_path_buf());
        self.files = crate::core::folder::scan_folder(folder, self.sort);
        self.last_folder = self.current_folder.clone();

        if let Some(target) = target_file {
            if let Some(pos) = self.files.iter().position(|f| f == target) {
                self.index = pos;
            } else {
                self.index = 0;
            }
        } else {
            self.index = 0;
        }
    }

    pub fn navigate_next(&mut self) {
        if !self.files.is_empty() && self.index < self.files.len() - 1 {
            self.index += 1;
            self.reset_zoom();
        }
    }

    pub fn navigate_prev(&mut self) {
        if self.index > 0 {
            self.index -= 1;
            self.reset_zoom();
        }
    }

    pub fn navigate_to(&mut self, index: usize) {
        if !self.files.is_empty() {
            self.index = index.min(self.files.len() - 1);
            self.reset_zoom();
        }
    }

    /// Each navigation starts the new image at fit-to-window, so a stuck zoom
    /// or unsaved rotate from a previous image can't bleed across.
    fn reset_zoom(&mut self) {
        self.zoom_fit = true;
        self.zoom = 1.0;
        self.pending_rotation = 0;
    }

    pub fn add_rotation(&mut self, degrees: u32) {
        self.pending_rotation = (self.pending_rotation + degrees) % 360;
    }

    pub fn cycle_sort(&mut self) {
        self.sort = self.sort.next();
    }

    /// Returns true if the mode changed.
    pub fn set_sort(&mut self, sort: SortMode) -> bool {
        if self.sort == sort {
            return false;
        }
        self.sort = sort;
        true
    }

    pub fn remove_file(&mut self, path: &Path) {
        if let Some(pos) = self.files.iter().position(|f| f == path) {
            self.files.remove(pos);
            if self.index >= self.files.len() {
                self.index = self.files.len().saturating_sub(1);
            } else if self.index > pos {
                self.index -= 1;
            }
        }
    }

    pub fn save(&self) {
        std::fs::create_dir_all(&self.config_dir).ok();
        let state_path = self.config_dir.join("state.toml");
        let mut content = String::new();
        if let Some(ref folder) = self.last_folder {
            content.push_str(&format!("last_folder = \"{}\"\n", folder.display()));
        }
        content.push_str(&format!("sort = \"{}\"\n", self.sort.as_str()));
        std::fs::write(state_path, content).ok();
    }

    pub fn restore(&mut self) {
        let state_path = self.config_dir.join("state.toml");
        if let Ok(content) = std::fs::read_to_string(&state_path) {
            if let Ok(table) = content.parse::<toml::Table>() {
                if let Some(toml::Value::String(s)) = table.get("last_folder") {
                    self.last_folder = Some(PathBuf::from(s));
                }
                if let Some(toml::Value::String(s)) = table.get("sort") {
                    self.sort = SortMode::parse(s);
                }
            }
        }
    }
}
